include/net/mqtt.h

```diff
@@ -499,6 +499,9 @@ struct mqtt_client {
 	/** MQTT protocol version. */
 	u8_t protocol_version;
 
+	/** Unanswered PINGREQ count on this connection. */
+	s8_t unacked_ping;
```

Comment:
is a signed byte always going to be enough? 
Is this specified by mqtt or how did you decide for it then?

Reply:
I think the idea was to use signed so that we notice underflow issue, the code checks that.

Reply:
Sure, but what about the case (if it's a relevant one, thus why I am asking) of overflow?

Reply:
According to the code, we just send PING and increment the counter. If we overflow the counter (==we have not received any reply), we could perhaps print something to the console and reset the value to 0. The MQTT library does not send PING itself, it is the application that needs to call this. So we only overflow if the connection is down and application just keeps calling mqtt_ping(), not sure how likely this is.

Reply:
documenting the ping limit in mqtt_ping() I guess

Reply:
There is no upper limit to the number of outstanding pings in the MQTT spec as far as I am aware.

In normal use cases, it's hard to get an outstanding ping count above single digit numbers. The only cases I can think of were the application were ping-flooding the server (bad behavior), or the TCP socket lost some traffic and was retrying but hadn't timed out yet and the application kept sending pings to check if it had recovered (more likely).

But even in the broken socket case, you can just send a single ping and wait for the count to return to 0.  Since it's TCP-based, all traffic prior to the ping has to be acknowledged in both directions before the application layer will receive the ACK and decrement the count.  Just sending a single ping and waiting for the decrement acts like a TCP sanity check and flush, all in one.

I actually went back and forth whether I should make it a larger integer just in case, but realized I can't think of any (realistic) use case where it needs to support values that big. On the other hand, the cost of making it 16 or 32 bit is pretty minimal, so I'm not opposed to doing it to be safe. I'll add an overflow check in the other direction even if we don't make it larger.  I think it should clamp at the max value though, and not reset to 0.

Reply:
I think s8_t is enough here. If you can submit a version that adds overflow check, then we can merge this one.

Reply:
+1